libs/base/System/Signal.idr

```diff
@@ -0,0 +1,259 @@
+data SignalError = Error Int
+
+getError : HasIO io => io SignalError
+getError = Error <$> primIO prim__getErrorNo
+
+isError : Int -> Bool
+isError (-1) = True
+isError _    = False
+
+||| Ignore the given signal.
+||| Be careful doing this, as most signals have useful
+||| default behavior -- you might want to set the signal
+||| to its default behavior instead with `defaultSignal`.
+export
+ignoreSignal : HasIO io => Signal -> io (Either SignalError ())
+ignoreSignal sig = do
+  res <- primIO $ prim__ignoreSignal (signalCode sig)
+  case isError res of
+       False => pure $ Right ()
+       True  => Left <$> getError
+
+||| Use the default handler for the given signal.
+||| You can use this function to unset custom
+||| handling of a signal.
+export
+defaultSignal : HasIO io => Signal -> io (Either SignalError ())
+defaultSignal sig = do
+  res <- primIO $ prim__defaultSignal (signalCode sig)
+  case isError res of
+       False => pure $ Right ()
+       True  => Left <$> getError
+
+||| Collect the given signal.
+|||
+||| This replaces the existing handling of the given signal
+||| and instead results in Idris collecting occurrences of
+||| the signal until you call `handleNextCollectedSignal`.
+|||
+||| First, call `collectSignal` for any number of signals.
+||| Then, call `handleNextCollectedSignal` in each main loop
+||| of your program to retrieve (and mark as handled) the next
+||| signal that was collected, if any.
+|||
+||| Multiple signals will be collected and can then be retrieved
+||| in the order they were received by multiple calls to
+||| `handleNextCollectedSignal`.
+|||
+||| You can call `handleManyCollectedSignals` to get a List of
+||| pending signals instead of retrieving them one at a time.
+export
+collectSignal : HasIO io => Signal -> io (Either SignalError ())
+collectSignal sig = do
+  res <- primIO $ prim__collectSignal (signalCode sig)
+  case isError res of
+       False => pure $ Right ()
+       True  => Left <$> getError
+
+||| Get next collected signal under the pretense of handling it.
+|||
+||| Calling this "marks" the signal as handled so the next time
+||| this function is called you will retrieve the next unhandled
+||| signal instead of the same signal again.
+|||
+||| You get back Nothing if there are no pending signals.
+export
+handleNextCollectedSignal : HasIO io => io (Maybe Signal)
+handleNextCollectedSignal =
+  toSignal <$> primIO prim__handleNextCollectedSignal
+
+||| Get many collected signals and mark them as handled.
+|||
+||| Use `forever` as your fuel if you don't want or need to
+||| retain totality. Alternatively, pick a max number to
+||| retrieve and use `limit/1` as your fuel.
+export
+handleManyCollectedSignals : HasIO io => Fuel -> io (List Signal)
+handleManyCollectedSignals fuel = catMaybes <$> handleMany fuel
+  where
+    handleMany : Fuel -> io (List (Maybe Signal))
+    handleMany Dry = pure []
+    handleMany (More fuel) = [| handleNextCollectedSignal :: handleMany fuel |]
```

Comment:
Shouldn't this stop at the first `Nothing`? This way if you don't know how many signals
you got but are willing to bet it's a finite number, `handleManyCollectedSignals forever`
would give you a finite list back rather than diverging.

Reply:
Yeah, definitely did not want this to diverge. I was blinded by how clean it looked in those brackets.

I've fixed, but if you've got a suggestion to make my fix any tidier, let me know! I don't find my fix _ugly_, but it does do a bit more explicit wrapping/unwrapping than my (admittedly broken) first attempt.

Reply:
Thanks! I don't plan on adding `unfoldM` for `Fuel` or tweaking this function more before merge, but I always learn a lot and appreciate you taking the time to answer.